File: lab_3_4.rb
def main
  # Чтение входных данных
  n, m = gets.split.map(&:to_i)
  nums = gets.split.map(&:to_i)

  # Проверка ограничений
  raise "Invalid n" unless (5..50000).cover?(n)
  raise "Invalid m" unless (4..100).cover?(m)
  raise "Invalid nums size" unless nums.size == n

  # Вычисление префиксных сумм
  prefix = Array.new(n + 1, 0)
  (0...n).each { |i| prefix[i+1] = prefix[i] + nums[i] }

  # Инициализация массива dp
  dp = Array.new(n + 1, -Float::INFINITY)
  dp[n] = 0 # Базовый случай

  # Заполнение массива dp
  (n-1).downto(0) do |i|
    max_k = [m, n - i].min
    (1..max_k).each do |k|
      current_sum = prefix[i + k] - prefix[i]
      dp[i] = [dp[i], current_sum - dp[i + k]].max
    end
  end

  # Определение победителя
  puts dp[0] > 0 ? 1 : 0
end

main

File: lab_3_4.cpp
// лаба 3.4

#include <iostream>
#include <vector>
#include <climits>

using namespace std;

/**
 * @file game_winner.cpp
 * @brief Определяет победителя в игре между Павлом и Викой.
 * 
 * Входные данные:
 * - n, m: целые числа (5 ≤ n ≤ 50000, 4 ≤ m ≤ 100).
 * - Последовательность из n целых чисел.
 * 
 * Выходные данные:
 * -1, если побеждает Павел; 0, если побеждает Вика.
 * 
 * Алгоритм:
 * - Использует динамическое программирование для вычисления оптимальной стратегии.
 * - Префиксные суммы ускоряют расчёт сумм подмассивов.
 * - dp[i] — максимальная разница очков, которую может получить игрок, начиная с позиции i.
 */

int main() {

    int n, m;
    cin >> n >> m;
    vector<int> nums(n);
    for (int i = 0; i < n; ++i) {
        cin >> nums[i];
    }
    
    // Префиксные суммы
    vector<long long> prefix(n + 1, 0);
    for (int i = 0; i < n; ++i) {
        prefix[i+1] = prefix[i] + nums[i];
    }
    
    // Массив dp
    vector<long long> dp(n + 1, LLONG_MIN);
    dp[n] = 0; // База
    
    for (int i = n-1; i >= 0; --i) {
        int max_k = min(m, n - i);
        for (int k = 1; k <= max_k; ++k) {
            long long current_sum = prefix[i + k] - prefix[i];
            dp[i] = max(dp[i], current_sum - dp[i + k]);
        }
    }
    
    cout << (dp[0] > 0 ? 1 : 0) << endl;
    return 0;
}
